firewood/src/merkle/changes.rs

```diff
@@ -0,0 +1,412 @@
+// Copyright (C) 2025, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE.md for licensing terms.
+
+use std::iter::once;
+
+use firewood_storage::{
+    Child, FileIoError, Node, Path, SharedNode, TrieHash, TrieReader, firewood_counter,
+};
+
+use crate::{
+    iter::key_from_nibble_iter,
+    merkle::{Key, PrefixOverlap},
+};
+
+/// Contains all of a node's state that is needed for node comparison in `DiffMerkleNodeStream`.
+/// It includes the nodes full path and its hash if available. `NodeState` is created from
+/// a `TraversalStackFrame` and differs from it by storing the node's full path instead of
+/// just its pre-path (doesn't include the node's partial path), and by storing the node as
+/// a `SharedNode` instead of a `Child`.
+#[derive(Clone)]
+struct NodeState {
+    path: Path,
+    node: SharedNode,
+    hash: Option<TrieHash>,
+}
+
+impl NodeState {
+    // Creates a NodeState from a `TraversalStackFrame` popped off the traversal stack, and a
+    // trie for reading the node from storage.
+    fn new(frame: TraversalStackFrame, trie: &dyn TrieReader) -> Result<Self, FileIoError> {
+        // If a hash is provided, then use that. Otherwise use the hash from the node.
+        let hash = frame
+            .hash
+            .or_else(|| frame.node.hash().map(|hash| hash.clone().into_triehash()));
+        let node = frame.node.as_shared_node(trie)?;
+        // Compute the full path for the node. This is used to compare with the full path
+        // of the current node from the other trie.
+        let mut path = frame.pre_path;
+        path.extend(node.partial_path().iter().copied());
+
+        Ok(Self { path, node, hash })
+    }
+}
+
+/// Contents of a frame on the traversal stack. Unlike for a `NodeState`, a `TraversalStackFrame`
+/// for a child can be generated from a parent without needing to load the child from storage.
+struct TraversalStackFrame {
+    pre_path: Path,
+    node: Child,
+    hash: Option<TrieHash>,
+}
+
+/// Contains the state required for performing pre-order iteration of a Merkle trie. This includes
+/// the state of the current node, a reference to a trie that implements the `TrieReader` trait,
+/// and a stack that contains nodes (and their associated state including its pre-path and hash) to
+/// be traversed with calls to `next`.
+struct PreOrderIterator<'a> {
+    node_state: Option<NodeState>,
+    trie: &'a dyn TrieReader,
+    traversal_stack: Vec<TraversalStackFrame>,
+}
+
+impl<'a> PreOrderIterator<'a> {
+    /// Create a pre-order iterator for the trie that starts at `start_key`. The iterator takes an
+    /// optional root hash which is available when the trie is from an `ImmutableProposal`.
+    #[cfg_attr(not(test), expect(dead_code))]
+    fn new(
+        trie: &'a dyn TrieReader,
```

Comment:
`dyn` references aren't free. They eliminate monomorphization and force all method calls to be done via dynamic dispatch. I don't particularly see a reason why we'd need to use a dyn reference here.